### DISCUSSION TITLE
Is there a simple way to replace an assembly reference without a full import?

### DISCUSSION THREAD

**Discussion by Serg046** [jbevain/cecil#732] [Q&A] [answered]
I have two assemblies A.dll and B.dll. I read A.dll, change its methods and then save it. One method "TestA()" has a reference to the method "TestB()" from B.dll and the thing is that I need to change some stuff inside "TestB()". If I just changed it through A.dll and then saved, my changed wouldn't be applied, the old B.dll would be in use. My idea is to save new B_patched.dll and override Domain.AssemblyResolve so that it loads B_patched.dll instead of B.dll. But if I didn't replace an assembly reference inside TestA() from B.dll to B_patched.dll, it wouldn't work, it would still use B.dll.

I understand that I could import the right method using `ImportReference(MethodDefinition)` but it is too complex and inefficient. The question is if I can replace a reference using a way like this:

```
var testA = asmA.Types[X].Methods[Y];
var testBReference = testA.Body.Instructions[Z].Operand as MethodReference;
testBReference.Module.Name += "_patched";
```

**Comment by jbevain** [comment, marked as answer]
I'm not 100% sure on what you're doing, but if I understand correctly the easiest solution in B.dll is to rename directly the assembly reference `module.AssemblyReferences[x].Name += " _patched"`, this way all the references like types and methods will be automatically scoped to the `_patched` assembly reference.

**Reply by Serg046** [reply]
This is what I tried initially and I thought for some reason that it didn't work. However the following example works fine:

```csharp
static void Main(string[] args)
{
	var consoleAsmName = typeof(Console).Assembly.GetName().Name;
	Console.WriteLine(consoleAsmName);
	var asmDef = AssemblyDefinition.ReadAssembly(Assembly.GetExecutingAssembly().Location);

	var consoleAsmRef = asmDef.MainModule.AssemblyReferences.Single(a => a.Name == consoleAsmName);
	consoleAsmRef.Name += "_patched";

	using var asmStream = new MemoryStream();
	asmDef.Write(asmStream);
	var asm = Assembly.Load(asmStream.ToArray());
	var type = asm.GetType(typeof(Program).FullName);
	type.InvokeMember(nameof(Print), BindingFlags.InvokeMethod, null, null, null);
}

public static void Print() => Console.WriteLine("Shouldn't be called");
```
The result is `FileNotFoundException` as expected, thank you!

**Reply by Serg046** [reply]
So now I can generate other assemblies and override them using the way like this:
```csharp
const string patched = "_patched";
var consoleAsm = typeof(Console).Assembly;
var consoleAsmName = consoleAsm.GetName().Name;
AppDomain.CurrentDomain.AssemblyResolve += (sender, eventArgs) =>
{
	if (eventArgs.Name.StartsWith(consoleAsmName + patched))
	{
		return consoleAsm;
	}
	throw new FileNotFoundException($"Cannot find {eventArgs.Name}");
};
```

**Reply by jbevain** [reply]
Great!
